### ISSUE TITLE
MITM possible with authenticated TCP transport noise stream?

### ISSUE THREAD

**Issue by ztane** [aholstenson/ataraxia#27]
I wrote a challenge authorization based `SharedSecretAuth`, but with secp256k1 public/private keys and trusted signatories... but then I noticed that if `SharedSecretAuth` or this new one I wrote one is used with plain `TCPTransport` then `context.remotePublicSecurity` and `context.localPublicSecurity` are `undefined` in the flows. Doesn't this mean that there in fact can be a man in the middle, and shouldn't the Noise public key hashes for example be available in the context to mitigate this?

**Comment by aholstenson** [comment]
Sorry for the late reply here. `localPublicSecurity` and `remotePublicSecurity` is something I had in the previous version of the transports to protect against MITM. 

I think the switch to using `noise-secret-stream` with an `XX` handshake does protect against MITM, but I might be wrong here. Looking at the pattern in Noise Explorer](https://noiseexplorer.com/patterns/XX/):

> Message D/E, sent by the initiator, benefits from sender and receiver authentication and is resistant to Key Compromise Impersonation. Assuming the corresponding private keys are secure, this authentication cannot be forged. Message contents benefit from message secrecy and strong forward secrecy: if the ephemeral private keys are secure and the responder is not being actively impersonated by an active attacker, message contents cannot be decrypted. 

If the use of the Noise handshake does protect against this the best way forward is probably to remove them, but if not the public key hashes should definitely go in there.

**Comment by ztane** [comment]
So the thing is there can not be any eavesdropper as a MITM listening the connection between the two noise-stream ends. However if the key pairs at both ends are ephemeral as they seem to be, then the case can be this:

```
A <===========> M <===========> B
  <-- Noise -->   <-- Noise -->

  Challenge
  ---------------------------->
                 
  Response/Challenge
  <----------------------------  
              
  Response
  ---------------------------->
```

Noise protects against key compromise impersonation i.e. that the *public key* of the other end is *that* public key, and there can be neither Eve nor Mallory in between. But in case of ephemeral keys Mallory can create its *own* keys and relay all messages between Alice and Bob and if the challenge-response protocol does not take this into account then it authenticates Alice and Bob whereas Noise ensures that the streams between Alice and Mallory, and Mallory and Bob are most secure.

I believe it would be sufficient to have local and remote public security be for example hashes of the `localPublicKey/remotePublicKey` and `remotePublicKey/localPublicKey` respectively for the challenge-response protocol - then Mallory cannot intervene.

**Comment by ztane** [comment]
if the challenge-response is integrated *inside* the noise stream handshake then it is fine but it was a bit difficult to see if this is the case

**Comment by ztane** [comment]
i.e. it is the case of `A` contacting `B'` and authenticating `B` -  the authentication of `B` must be bound to the keys of `B'` to ensure that `B'` *is* `B` too.

**Comment by aholstenson** [comment]
Sorry for the delay here. I've finally found the time to add back the support for this to the Noise connections.

**Comment by aholstenson** [comment]
Released in 0.12.0 - feel free to reopen this if there's any issues

**Comment by ztane** [comment]
I'll check. Thanks!
